src/StreamFactoryInterface.php

```diff
@@ -9,15 +9,11 @@
     /**
      * Create a new stream from a resource or a string.
      *
-     * If the argument is a resource, it MUST be readable and SHOULD be seekable. It MAY be writable.
-     * If the argument is a string, a temporary resource will be created that is writable and seekable.
-     * 
-     * If the arugment is a string it will be interpided as the content of the stream. File names or 
-     * file paths are not supported. 
+     * String content will be used as the body of the stream. Files paths are not supported.
```

Comment:
> It is perfectly acceptable to always rely on php://temp without worrying about memory/fs issues.

Okay, valid - using `php://memory` is a very rare use-case and therefore unimportant in terms of interop. Agreed.

> Besides supposed purity, what benefit would having multiple streams have for package authors and end users?

My concern isn't purity, but scope.

The main benefit of separation, in my opinion, is that the following two interfaces are simpler, easier to describe, and easier to understand.

To the point - and pay attention to the inline documentation - compare these:

```php
interface FileStreamFactoryInterface {
    /**
     * Opens a file and creates a Stream object.
     *
     * @param string $path absolute path of the file to open
     * @param string $mode file mode, as per fopen()
     *
     * @return StreamInterface
     */
    public function createFileStream(string $path, string $mode = "r");
}

interface ResourceStreamFactoryInterface {
    /**
     * Creates a Stream object for a given `resource` handle.
     *
     * @param resource $handle
     *
     * @return StreamInterface
     */
    public function createResourceStream($handle);
}
```

Now compare with this:

```php
interface StreamFactoryInterface {
    /**
     * Creates a Stream object for a given `resource` handle, or
     * opens a file at a given path and creates a Stream object.
     *
     * @param resource|string $resource_or_path a resource handle, or an absolute
     *                                          path of a file to open.
     * @param string $mode file mode, as per fopen()
     *                     (if the first argument is a resource, this is unused)
     *
     * @return StreamInterface
     */
    public function createStream($resource_or_path, $mode = "r");
}
```

The explanations require more "or" and "if", and you have to explain that the `$mode` argument may actually have no effect.

Using this overloaded interface may be *easier*, once you understand how it works, but the non-overloaded interfaces are *simpler*, and therefore easier to explain and understand individually.

PHP does not formally support overloading, and php-doc does not define a means of formally documenting an overloaded interface - so the descriptions have to be written in english, which means that static analysis will be wrong; an IDE, for example, will auto-complete for the second argument, even though, technically, the second argument does not "exist" if the first argument is a `resource`, so this does have implications for authors and end-users.

With that said, maybe the need to specify the `$mode` is not a common enough use-case? I don't have much experience with Guzzle and client-middleware, so I can't say if read/write-streams are important in terms of interop. Thoughts?

If `$mode` was not required, a single overloaded interface would at least be consistent by number of arguments, and the description somewhat simpler:

```php
interface StreamFactoryInterface {
    /**
     * Creates a Stream object for a given `resource` handle; or opens a
     * file at a given path, in read-only mode, and creates a Stream object.
     *
     * @param resource|string $resource_or_path a resource handle, or an absolute
     *                                          path of a file to open.
     *
     * @return StreamInterface
     */
    public function createStream($resource_or_path);
}
```

I would still have reservations about the interface and method names though - when I see a non-descript factory name like `createStream()`, I immediately think "create stream how?", and "what type of stream?", especially if I've just seen `createTempStream()` elsewhere.

If we go that route, I'd prefer qualifying the interface/method-names as e.g. `ResourceStreamFactoryInterface::createResourceStream()`, to clarify the fact that this factory always works with a `resource` and may implicitly open one.

I would still strongly prefer factory interface with one responsibility, but I could bend on this, if you're sure that opening files in read-only mode is the most important thing in terms of interop? E.g. read/write-mode is a rare case. Is it?